<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-04-27 Mon 16:25 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>AD brain-specific TWAS analysis</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css"/>
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2019 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">AD brain-specific TWAS analysis</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org2c5c243">Introduction</a></li>
<li><a href="#org693dfd1">Setup</a></li>
<li><a href="#org679ac45">Data</a>
<ul>
<li><a href="#org5b2490c">Fitted elastic net models</a></li>
<li><a href="#org94e4a19">UTMOST summary statistics</a></li>
</ul>
</li>
<li><a href="#orga2da95b">Results</a>
<ul>
<li><a href="#org2bcbedd">sqtl associations</a></li>
<li><a href="#orgc9b0e49">fqtl associations</a></li>
<li><a href="#orge074c25">UTMOST associations</a></li>
<li><a href="#org10118b4">Comparison</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org2c5c243" class="outline-2">
<h2 id="org2c5c243">Introduction</h2>
<div class="outline-text-2" id="text-org2c5c243">
<p>
Here, we perform discovery and replication of AD gene-level associations in
brain tissues using <code>fqtl</code> models.
</p>
</div>
</div>

<div id="outline-container-org693dfd1" class="outline-2">
<h2 id="org693dfd1">Setup</h2>
<div class="outline-text-2" id="text-org693dfd1">
<div class="org-src-container">
<pre class="src src-ipython" id="org0501ad2"><span class="org-keyword">import</span> glob
<span class="org-keyword">import</span> itertools <span class="org-keyword">as</span> it
<span class="org-keyword">import</span> os.path
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scipy.linalg <span class="org-keyword">as</span> sl
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> statsmodels.api <span class="org-keyword">as</span> sm
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org679ac45" class="outline-2">
<h2 id="org679ac45">Data</h2>
<div class="outline-text-2" id="text-org679ac45">
</div>
<div id="outline-container-org5b2490c" class="outline-3">
<h3 id="org5b2490c">Fitted elastic net models</h3>
<div class="outline-text-3" id="text-org5b2490c">
<p>
Gamazon et al. 2015 trained elastic net models on GTEx v6p. Download the
fitted models from PredictDB.
</p>

<div class="org-src-container">
<pre class="src src-shell">curl -OL <span class="org-string">"https://zenodo.org/record/3572842/files/GTEx-V6p-1KG-2016-11-16.tar.gz"</span>
tar xf GTEx-V6p-1KG-2016-11-16.tar.gz
rm GTEx-V6p-1KG-2016-11-16.tar.gz
</pre>
</div>

<p>
The database uses rsid as the key for the SNPs. Download the dbSNP
annotation to convert rsids to (chromosome, start, end).
</p>

<div class="org-src-container">
<pre class="src src-shell">curl -O --ftp-pasv <span class="org-string">"https://ftp.ncbi.nih.gov/snp/pre_build152/organisms/human_9606_b151_GRCh37p13/VCF/00-common_all.vcf.gz"</span> 
curl -O --ftp-pasv <span class="org-string">"https://ftp.ncbi.nih.gov/snp/pre_build152/organisms/human_9606_b151_GRCh37p13/VCF/00-common_all.vcf.gz.tbi"</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org94e4a19" class="outline-3">
<h3 id="org94e4a19">UTMOST summary statistics</h3>
<div class="outline-text-3" id="text-org94e4a19">
<p>
Download <i>UTMOST</i> summary statistics.
</p>

<div class="org-src-container">
<pre class="src src-shell">curl --ftp-pasv -OL <span class="org-string">"ftp://ftp.biostat.wisc.edu/pub/lu_group/Projects/UTMOST/load_utmost_sumstats/utmost.cross.tissue.sumstats.igap.discovery.txt"</span>
curl --ftp-pasv -OL <span class="org-string">"ftp://ftp.biostat.wisc.edu/pub/lu_group/Projects/UTMOST/load_utmost_sumstats/utmost.cross.tissue.sumstats.gwax.replicate.txt"</span>
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orga2da95b" class="outline-2">
<h2 id="orga2da95b">Results</h2>
<div class="outline-text-2" id="text-orga2da95b">
</div>
<div id="outline-container-org2bcbedd" class="outline-3">
<h3 id="org2bcbedd">sqtl associations</h3>
<div class="outline-text-3" id="text-org2bcbedd">
<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">n_chunks</span> = 29
<span class="org-variable-name">igap_sqtl_stat</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/imputed_IGAP_Alzheimer-{i}.sqtl.stat.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)])
<span class="org-variable-name">igap_sqtl_pval</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/imputed_IGAP_Alzheimer-{i}.sqtl.pval.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)])
<span class="org-variable-name">gwax_sqtl_stat</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/ad-gwax-hg37-{i}.sqtl.stat.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)])
<span class="org-variable-name">gwax_sqtl_pval</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/ad-gwax-hg37-{i}.sqtl.pval.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)])
</pre>
</div>

<p>
Find the significant associations in the discovery cohort (IGAP; FDR 5%).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">igap_sqtl_sig</span> = igap_sqtl_pval.<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: sm.stats.multipletests(x, alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0])
</pre>
</div>

<p>
For the significant associations, determine whether they replicated (GWAX;
FDR 5%).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gwax_sqtl_rep</span> = gwax_sqtl_pval[igap_sqtl_sig].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: sm.stats.multipletests(x, alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0])
</pre>
</div>
</div>
</div>

<div id="outline-container-orgc9b0e49" class="outline-3">
<h3 id="orgc9b0e49">fqtl associations</h3>
<div class="outline-text-3" id="text-orgc9b0e49">
<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">n_chunks</span> = 29
<span class="org-variable-name">igap_fqtl_stat</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/imputed_IGAP_Alzheimer-{i}.fqtl.stat.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)], sort=<span class="org-constant">True</span>)
<span class="org-variable-name">igap_fqtl_pval</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/imputed_IGAP_Alzheimer-{i}.fqtl.pval.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)], sort=<span class="org-constant">True</span>)
<span class="org-variable-name">gwax_fqtl_stat</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/ad-gwax-hg37-{i}.fqtl.stat.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)], sort=<span class="org-constant">True</span>)
<span class="org-variable-name">gwax_fqtl_pval</span> = pd.concat([pd.read_csv(f<span class="org-string">'/broad/hptmp/aksarkar/twas/ad-gwax-hg37-{i}.fqtl.pval.txt.gz'</span>, index_col=0, sep=<span class="org-string">'\t'</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_chunks)], sort=<span class="org-constant">True</span>)
</pre>
</div>

<p>
Find the significant associations in the discovery cohort (IGAP; FDR 5%),
and determine whether they replicated (GWAX; FDR 5%).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">igap_fqtl_sig</span> = igap_fqtl_pval.<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: sm.stats.multipletests(x, alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0])
<span class="org-variable-name">gwax_fqtl_rep</span> = gwax_fqtl_pval[igap_fqtl_sig].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: sm.stats.multipletests(x, alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0])
</pre>
</div>
</div>
</div>

<div id="outline-container-orge074c25" class="outline-3">
<h3 id="orge074c25">UTMOST associations</h3>
<div class="outline-text-3" id="text-orge074c25">
<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">igap_utmost_pval</span> = pd.read_csv(<span class="org-string">'/broad/hptmp/aksarkar/utmost.cross.tissue.sumstats.igap.discovery.txt'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
<span class="org-variable-name">gwax_utmost_pval</span> = pd.read_csv(<span class="org-string">'/broad/hptmp/aksarkar/utmost.cross.tissue.sumstats.gwax.replicate.txt'</span>, sep=<span class="org-string">'\t'</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">igap_utmost_sig</span> = sm.stats.multipletests(igap_utmost_pval[<span class="org-string">'pvalue'</span>], alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0]
<span class="org-variable-name">gwax_utmost_rep</span> = sm.stats.multipletests(gwax_utmost_pval.loc[igap_utmost_sig, <span class="org-string">'pvalue'</span>], alpha=0.1, method=<span class="org-string">'fdr_bh'</span>)[0]
</pre>
</div>
</div>
</div>

<div id="outline-container-org10118b4" class="outline-3">
<h3 id="org10118b4">Comparison</h3>
<div class="outline-text-3" id="text-org10118b4">
<p>
Compare the number of discovered/replicated associations in brain tissues
between sqtl/fqtl.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">grid</span> = np.arange(gwax_fqtl_rep.shape[1])
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(7.5, 5.5)

ax[0].bar(grid, igap_sqtl_sig.<span class="org-builtin">sum</span>(), color=<span class="org-string">'k'</span>, label=<span class="org-string">'sqtl'</span>)
ax[0].bar(grid, igap_fqtl_sig.<span class="org-builtin">sum</span>(), color=<span class="org-string">'r'</span>, alpha=0.5, label=<span class="org-string">'fqtl'</span>)
ax[0].legend()
ax[0].set_ylabel(<span class="org-string">'Num disc assoc'</span>)

ax[1].bar(grid, gwax_sqtl_rep.<span class="org-builtin">sum</span>() / igap_sqtl_sig.<span class="org-builtin">sum</span>(), color=<span class="org-string">'k'</span>)
ax[1].bar(grid, gwax_fqtl_rep.<span class="org-builtin">sum</span>() / igap_fqtl_sig.<span class="org-builtin">sum</span>(), color=<span class="org-string">'r'</span>, alpha=0.5)
ax[1].set_xticks(grid)
ax[1].set_xticklabels(gwax_fqtl_rep.columns, rotation=90)
ax[1].set_xlabel(<span class="org-string">'Tissue'</span>)
ax[1].set_ylabel(<span class="org-string">'Prop repl'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/brain.org/sqtl-fqtl-ad-repl.png" alt="sqtl-fqtl-ad-repl.png">
</p>
</div>

<p>
Compare the total number of associations (across tissues) between sqtl,
fqtl, and UTMOST.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.Series({<span class="org-string">'sqtl'</span>: igap_sqtl_sig.agg(<span class="org-builtin">any</span>, axis=1).<span class="org-builtin">sum</span>(),
           <span class="org-string">'fqtl'</span>: igap_fqtl_sig.agg(<span class="org-builtin">any</span>, axis=1).<span class="org-builtin">sum</span>(),
           <span class="org-string">'utmost'</span>: igap_utmost_sig.<span class="org-builtin">sum</span>()})
</pre>
</div>

<pre class="example">
sqtl       73
fqtl       37
utmost    192
dtype: int64
</pre>

<p>
Compare the total number of replications.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.Series({<span class="org-string">'sqtl'</span>: gwax_sqtl_rep.agg(<span class="org-builtin">any</span>, axis=1).<span class="org-builtin">sum</span>(),
           <span class="org-string">'fqtl'</span>: gwax_fqtl_rep.agg(<span class="org-builtin">any</span>, axis=1).<span class="org-builtin">sum</span>(),
           <span class="org-string">'utmost'</span>: gwax_utmost_rep.<span class="org-builtin">sum</span>()})
</pre>
</div>

<pre class="example">
sqtl      26
fqtl      10
utmost    29
dtype: int64
</pre>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-04-27 Mon 16:25</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
