<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-04-27 Mon 16:17 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>TWAS pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css"/>
<link rel="stylesheet" type="text/css" href="css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="css/main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2019 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">TWAS pipeline</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orged0fb56">Introduction</a></li>
<li><a href="#orgbb95c24">Setup</a></li>
<li><a href="#org574bc34">Data</a>
<ul>
<li><a href="#orgcafff56">Pre-process the summary statistics</a></li>
<li><a href="#orgb348d59">AD replication data</a></li>
<li><a href="#orga92d915">Prepare the genotype matrices</a></li>
</ul>
</li>
<li><a href="#results">Results</a>
<ul>
<li><a href="#orgc14368f">APOE example</a></li>
<li><a href="#org7afd361">Run TWAS</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orged0fb56" class="outline-2">
<h2 id="orged0fb56">Introduction</h2>
<div class="outline-text-2" id="text-orged0fb56">
<p>
Here, we use the fitted <code>fqtl</code> models to perform TWAS of 114 traits
(<a href="https://doi.org/10.1038/s41467-018-03621-1">Barbeira et al 2018</a>).\(
  \DeclareMathOperator\N{N}
  \newcommand\mU{\mathbf{U}}
  \newcommand\md{\mathbf{D}}
  \newcommand\mi{\mathbf{I}}
  \newcommand\mr{\mathbf{R}}
  \newcommand\mv{\mathbf{V}}
  \newcommand\mx{\mathbf{X}}
  \newcommand\vb{\mathbf{b}_t}
  \newcommand\veta{\boldsymbol{\eta}_t}
  \newcommand\vy{\mathbf{y}}
  \newcommand\vz{\mathbf{z}}
  \)
</p>
</div>
</div>

<div id="outline-container-orgbb95c24" class="outline-2">
<h2 id="orgbb95c24">Setup</h2>
<div class="outline-text-2" id="text-orgbb95c24">
<div class="org-src-container">
<pre class="src src-ipython" id="org7be5838"><span class="org-keyword">import</span> glob
<span class="org-keyword">import</span> itertools <span class="org-keyword">as</span> it
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> os.path
<span class="org-keyword">import</span> os.path
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> pyplink
<span class="org-keyword">import</span> scipy.linalg <span class="org-keyword">as</span> sl
<span class="org-keyword">import</span> scipy.sparse <span class="org-keyword">as</span> ss
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> sqlite3
<span class="org-keyword">import</span> tabix
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org574bc34" class="outline-2">
<h2 id="org574bc34">Data</h2>
<div class="outline-text-2" id="text-org574bc34">
</div>
<div id="outline-container-orgcafff56" class="outline-3">
<h3 id="orgcafff56">Pre-process the summary statistics</h3>
<div class="outline-text-3" id="text-orgcafff56">
<p>
The GTEx consortium provided summary statistics in hg38. <code>liftover</code> back to
hg19 to combine with v6p models.
</p>

<div class="org-src-container">
<pre class="src src-shell">find /broad/compbio/data/gtex_gwas_hg38_tabix/ -name <span class="org-string">"*.bed.gz"</span> &gt;manifest
</pre>
</div>

<div class="org-src-container">
<pre class="src src-shell">qsub -cwd -terse -V -sync n -j y -N liftover -S /bin/bash -t 36 -l <span class="org-variable-name">h_vmem</span>=4G
<span class="org-builtin">set</span> -e
<span class="org-keyword">function</span> <span class="org-function-name">lift</span> () {
    <span class="org-variable-name">in</span>=gtex-gwas-hg19/$(<span class="org-sh-quoted-exec">basename</span> $<span class="org-variable-name">1</span> .bed.gz).hg38
    zcat $<span class="org-variable-name">1</span> | awk -vOFS=<span class="org-string">'\t'</span> <span class="org-string">'NR &gt; 1 {print $1, $2, $3, $4"|"$5"|"$6"|"$7}'</span> &gt;$<span class="org-variable-name">in</span>
    <span class="org-variable-name">out</span>=gtex-gwas-hg19/$(<span class="org-sh-quoted-exec">basename</span> $<span class="org-variable-name">1</span> .bed.gz).hg19
    liftOver $<span class="org-variable-name">in</span> /broad/compbio/data/gtex_eqtl_tabix/hg38ToHg19.over.chain.gz $<span class="org-variable-name">out</span> $<span class="org-variable-name">out</span>.unmapped
    sort -k1,1 -k2,2n $<span class="org-variable-name">out</span> | tr <span class="org-string">'|'</span> <span class="org-string">'\t'</span> | bgzip &gt;gtex-gwas-hg19/$(<span class="org-sh-quoted-exec">basename</span> $<span class="org-variable-name">1</span>)
    tabix gtex-gwas-hg19/$(<span class="org-sh-quoted-exec">basename</span> $<span class="org-variable-name">1</span>)
}
<span class="org-builtin">readarray</span> -O1 tasks &lt;manifest
lift ${<span class="org-variable-name">tasks</span>[$<span class="org-variable-name">SGE_TASK_ID</span>]}
</pre>
</div>
</div>
</div>

<div id="outline-container-orgb348d59" class="outline-3">
<h3 id="orgb348d59">AD replication data</h3>
<div class="outline-text-3" id="text-orgb348d59">
<p>
Download ADGC phase 2 (<a href="https://www.niagads.org/datasets/ng00076">https://www.niagads.org/datasets/ng00076</a>)
</p>

<div class="org-src-container">
<pre class="src src-shell">curl -sL <span class="org-string">"https://www.niagads.org/system/tdf/public_docs/ADGC2_ModelB_METAL_COMMON.InvVar.results.formatted_p-value_only.txt?file=1&amp;type=field_collection_item&amp;id=100&amp;force="</span> -o adgc2.txt
gzip adgc2.txt
</pre>
</div>

<p>
The data is on hg19 (GHCr37).
</p>

<div class="org-src-container">
<pre class="src src-shell">zcat adgc2.txt.gz | awk <span class="org-string">'NR == 2 {split($1, a, ":"); print a[2]; exit}'</span> | xargs -I{} zgrep -wm1 {} /broad/compbio/lward/incoming/dbSNP/hg19_b137/bed_chr_1.txt.gz
</pre>
</div>

<p>
The publicly available data does not have \(z\)-scores or odds ratios.
</p>

<p>
Download GWAX summary statistics
(<a href="https://www.nature.com/articles/ng.3766">Liu et al. 2017</a>).
</p>

<div class="org-src-container">
<pre class="src src-shell">qsub -terse -cwd -V -j y -sync n
curl -sOL <span class="org-string">"http://gwas-browser.nygenome.org/downloads/gwas-browser/AD.gwax.assoc.gz"</span>
</pre>
</div>

<p>
The data is on hg19.
</p>

<div class="org-src-container">
<pre class="src src-shell">zgrep -wm1 rs144155419 /broad/compbio/lward/incoming/dbSNP/hg19_b137/bed_chr_1.txt.gz
</pre>
</div>

<p>
Convert \(p\)-values to \(z\)-scores and format the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">data</span> = pd.read_csv(<span class="org-string">'/broad/hptmp/aksarkar/AD.gwax.assoc.gz'</span>, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">data</span>[<span class="org-string">'chr'</span>] = data[<span class="org-string">'CHR'</span>].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: f<span class="org-string">'chr{x}'</span>)
<span class="org-variable-name">data</span>[<span class="org-string">'start'</span>] = data[<span class="org-string">'BP'</span>]
<span class="org-comment-delimiter"># </span><span class="org-comment">Important: we removed indels from GTEx</span>
<span class="org-variable-name">data</span>[<span class="org-string">'end'</span>] = data.<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x[<span class="org-string">'BP'</span>] <span class="org-keyword">if</span> <span class="org-builtin">len</span>(x[<span class="org-string">'A1'</span>]) == <span class="org-builtin">len</span>(x[<span class="org-string">'A2'</span>]) == 1 <span class="org-keyword">else</span> np.nan, axis=1).astype(<span class="org-string">'Int64'</span>)
<span class="org-variable-name">data</span>[<span class="org-string">'zscore'</span>] = np.sqrt(st.chi2(1).sf(data[<span class="org-string">'P'</span>])) * np.sign(np.log(data[<span class="org-string">'OR'</span>]))
<span class="org-variable-name">data</span>[<span class="org-string">'pval'</span>] = data[<span class="org-string">'P'</span>]
(data[[<span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'A1'</span>, <span class="org-string">'A2'</span>, <span class="org-string">'zscore'</span>, <span class="org-string">'pval'</span>]]
 .dropna()
 .to_csv(<span class="org-string">'/broad/hptmp/aksarkar/gtex-gwas-hg19/ad-gwax-hg37.bed'</span>, index=<span class="org-constant">None</span>, header=<span class="org-constant">None</span>, sep=<span class="org-string">'\t'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-shell">qsub -cwd -terse -V -sync n -j y -N tabix -S /bin/bash
<span class="org-builtin">set</span> -e
bgzip /broad/hptmp/aksarkar/gtex-gwas-hg19/ad-gwax-hg37.bed
tabix /broad/hptmp/aksarkar/gtex-gwas-hg19/ad-gwax-hg37.bed.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-orga92d915" class="outline-3">
<h3 id="orga92d915">Prepare the genotype matrices</h3>
<div class="outline-text-3" id="text-orga92d915">
<p>
Extract genotypes within 1 megabase of the TSS for each gene. Genes are
indexed in the manifest.
</p>

<div class="org-src-container">
<pre class="src src-shell">awk <span class="org-string">'$6 &gt;= 10'</span> /broad/compbio/aksarkar/projects/gtex-fqtl/data/fqtl-*-valid.genes.txt &gt;manifest
qsub -cwd -V -terse -sync n -j y -N plink -t 1-100 -l <span class="org-variable-name">h_vmem</span>=2G
awk -vn=$<span class="org-variable-name">SGE_TASK_LAST</span> -vi=$<span class="org-variable-name">SGE_TASK_ID</span> <span class="org-string">'NR % n == i - 1 {w = 1000000; ub = $4 + w; if ($4 &gt; w) {lb = $4 - w} else {lb = 0} system("plink --memory 2000 --bfile /broad/compbio/data/GTEx/GTEx_restricted/v6_plink/BED_qc/chr"$3" --make-bed --chr "$3" --from-bp "lb" --to-bp "ub" --out "$1)}'</span> manifest
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org7554c3a" class="outline-2">
<h2 id="results"><a id="org7554c3a"></a>Results</h2>
<div class="outline-text-2" id="text-results">
</div>
<div id="outline-container-orgc14368f" class="outline-3">
<h3 id="orgc14368f">APOE example</h3>
<div class="outline-text-3" id="text-orgc14368f">
<p>
Extract the <code>fqtl</code> model for <i>APOE</i>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_id</span> = 49282
<span class="org-variable-name">prefix</span> = f<span class="org-string">'/broad/compbio/ypp/gtex/analysis/fqtl-gtex/result/fqtl-std/{gene_id}/50'</span>
<span class="org-variable-name">snp_annot</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.snp.info.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">snp_lodds</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.snp.lodds.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">snp_lodds.index</span> = snp_annot[3]
<span class="org-variable-name">tis_annot</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.tis.info.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>, index_col=0)
<span class="org-variable-name">tis_lodds</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.tis.lodds.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">tis_lodds.index</span> = tis_annot[1].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.rsplit(<span class="org-string">'_'</span>, maxsplit=1)[0])
<span class="org-variable-name">keep</span> = (sp.expit(snp_lodds) &gt; .95).<span class="org-builtin">any</span>(axis=0)
<span class="org-variable-name">L</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.snp.theta.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">L.index</span> = snp_annot[3]
<span class="org-variable-name">F</span> = pd.read_csv(f<span class="org-string">'{prefix}/fqtl.tis.theta.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">F.index</span> = tis_annot[1].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x[:-9])
<span class="org-variable-name">B</span> = L.loc[:,keep].dot(F.loc[:,keep].T)
</pre>
</div>

<p>
Look at the fitted <code>fqtl</code> loadings and factors.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(7.5, 5)
<span class="org-variable-name">query</span> = F.loc[:,keep]
<span class="org-variable-name">grid</span> = np.arange(query.shape[0])
plt.axhline(y=0, lw=1, color=<span class="org-string">'0.7'</span>)
plt.grid(axis=<span class="org-string">'x'</span>, color=<span class="org-string">'0.7'</span>)
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(query):
  plt.bar(grid, query[k], width=0.9 , color=cm(i), label=f<span class="org-string">'Factor {i}'</span>, zorder=3)
plt.legend()
plt.xticks(grid, stat.index, rotation=90)
plt.xlabel(<span class="org-string">'Tissue'</span>)
plt.ylabel(<span class="org-string">'Factor weight'</span>)
plt.title(<span class="org-string">'APOE'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/twas.org/apoe-fqtl.png" alt="apoe-fqtl.png">
</p>
</div>

<p>
Extract the <code>sqtl</code> models for <i>APOE</i>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">chromosome</span> = 19
<span class="org-variable-name">prefix</span> = f<span class="org-string">'/broad/compbio/ypp/gtex/analysis/fqtl-gtex/result/sqtl/{chromosome}/{gene_id}'</span>
<span class="org-variable-name">sqtl_tis_annot</span> = pd.read_csv(f<span class="org-string">'{prefix}/sqtl.tis.info.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>, index_col=0)
<span class="org-variable-name">sqtl_snp_annot</span> = pd.read_csv(f<span class="org-string">'{prefix}/sqtl.snp.info.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">sqtl_snp_lodds</span> = pd.read_csv(f<span class="org-string">'{prefix}/sqtl.effect.lodds.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">sqtl_snp_lodds.index</span> = snp_annot[3]
<span class="org-variable-name">sqtl_snp_lodds.columns</span> = sqtl_tis_annot[1].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.rsplit(<span class="org-string">'_'</span>, maxsplit=1)[0])
<span class="org-variable-name">sqtl_B</span> = pd.read_csv(f<span class="org-string">'{prefix}/sqtl.effect.theta.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>)
<span class="org-variable-name">sqtl_B.index</span> = snp_annot[3]
<span class="org-variable-name">sqtl_B.columns</span> = sqtl_tis_annot[1].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.rsplit(<span class="org-string">'_'</span>, maxsplit=1)[0])
</pre>
</div>

<p>
Load the <i>cis</i>-genotypes around <i>APOE</i>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> pyplink.PyPlink(f<span class="org-string">'/broad/hptmp/aksarkar/geno/{gene_id}'</span>) <span class="org-keyword">as</span> f:
  <span class="org-variable-name">fam</span> = f.get_fam()
  <span class="org-variable-name">bim</span> = f.get_bim()
  <span class="org-variable-name">x</span> = np.ma.masked_equal(np.array([row <span class="org-keyword">for</span> _, row <span class="org-keyword">in</span> f.iter_geno()]).T, -1).astype(<span class="org-builtin">float</span>)
  <span class="org-variable-name">x</span> -= x.mean()
  <span class="org-variable-name">x</span> /= x.std()
  <span class="org-variable-name">x</span> = pd.DataFrame(x, index=fam[<span class="org-string">'fid'</span>], columns=bim[<span class="org-string">'pos'</span>])
</pre>
</div>

<p>
Read the GWAS summary statistics.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">genes</span> = pd.read_csv(<span class="org-string">'/broad/hptmp/aksarkar/geno/manifest'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>, index_col=0)
<span class="org-variable-name">f</span> = tabix.<span class="org-builtin">open</span>(<span class="org-string">'/broad/hptmp/aksarkar/gtex-gwas-hg19/imputed_IGAP_Alzheimer.bed.gz'</span>)
<span class="org-variable-name">row</span> = genes.loc[gene_id]
<span class="org-variable-name">gwas_z</span> = (pd.DataFrame(f.query(f<span class="org-string">'chr{row[2]}'</span>, <span class="org-builtin">int</span>(row[3] - 1e6), <span class="org-builtin">int</span>(row[3] + 1e6)))
          .astype(<span class="org-builtin">dict</span>(<span class="org-builtin">enumerate</span>([<span class="org-builtin">str</span>, <span class="org-builtin">int</span>, <span class="org-builtin">int</span>, <span class="org-builtin">str</span>, <span class="org-builtin">str</span>, <span class="org-builtin">float</span>, <span class="org-builtin">float</span>]))))
gwas_z.head()
</pre>
</div>

<pre class="example">
0         1         2  3  4       5       6
0  chr19  44409410  44409411  T  A  2.4210  0.0155
1  chr19  44410383  44410384  A  G -0.1969  0.8440
2  chr19  44410421  44410422  G  A -1.4283  0.1530
3  chr19  44410613  44410614  A  G -1.6997  0.0892
4  chr19  44410845  44410846  A  G -2.3529  0.0186
</pre>

<p>
Merge the tissue-specific effect sizes and the GWAS summary statistics.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">twas_data</span> = gwas_z.merge(snp_annot, left_on=[2, 3, 4], right_on=[3, 4, 5]).set_index(<span class="org-string">'2_x'</span>)
twas_data.head(n=1)
</pre>
</div>

<pre class="example">
2  3  4    0_x       1_x 3_x 4_x     5_x      6  0_y  \
2_x
44410422  44410422  G  A  chr19  44410421   G   A -1.4283  0.153   19

1_y  2_y       3_y 4_y 5_y
2_x
44410422  19_44410422_A_G_b37    0  44410422   G   A
</pre>

<p>
Plot the TWAS loadings and the GWAS \(z\)-scores.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = pd.read_csv(<span class="org-string">'/broad/hptmp/aksarkar/geno/manifest'</span>, sep=<span class="org-string">'\t'</span>, header=<span class="org-constant">None</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(3, 1, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(7, 4.5)

<span class="org-variable-name">query</span> = sqtl_B.loc[:,(sp.expit(sqtl_snp_lodds) &gt; .95).<span class="org-builtin">any</span>(axis=0)]
<span class="org-variable-name">pos</span> = query.index / 1e6
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(query):
  <span class="org-variable-name">keep_snps</span> = sp.expit(sqtl_snp_lodds[k]) &gt; .95
  ax[0].scatter(pos[keep_snps], query[k][keep_snps], s=4, color=cm(1 - i), label=k, zorder=3)

<span class="org-variable-name">query</span> = L.loc[:,keep].align(twas_data, axis=0, join=<span class="org-string">'inner'</span>)[0]
<span class="org-variable-name">query_lodds</span> = snp_lodds.loc[:,keep].align(twas_data, axis=0, join=<span class="org-string">'inner'</span>)[0]
<span class="org-variable-name">pos</span> = query.index / 1e6
<span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(query):
  <span class="org-variable-name">keep_snps</span> = query_lodds[k] &gt; sp.logit(.95)
  ax[1].scatter(pos[keep_snps], query[k][keep_snps], s=4, color=cm(i), label=f<span class="org-string">'Factor {i}'</span>, zorder=3)

ax[2].scatter(pos, twas_data[<span class="org-string">'5_x'</span>], s=1, color=<span class="org-string">'k'</span>, zorder=3, label=<span class="org-constant">None</span>)

<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  <span class="org-keyword">for</span> t <span class="org-keyword">in</span> pos[keep_snps]:
    a.axvline(t, lw=0.5, c=<span class="org-string">'0.85'</span>)
  a.axhline(y=0, lw=1, color=<span class="org-string">'0.7'</span>)
  a.axvline(gene_info.loc[gene_id,3] / 1e6, c=<span class="org-string">'b'</span>, lw=0.5)

ax[0].set_ylabel(<span class="org-string">'sQTL effect size'</span>)
ax[0].legend(frameon=<span class="org-constant">False</span>, markerscale=2, handletextpad=0, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
ax[1].set_ylabel(<span class="org-string">'fQTL loading'</span>)
ax[1].legend(frameon=<span class="org-constant">False</span>, markerscale=2, handletextpad=0, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
ax[2].set_ylabel(<span class="org-string">'AD GWAS $z$-score'</span>)
ax[1].set_xlabel(<span class="org-string">'Genomic position (MB)'</span>)
ax[0].set_title(<span class="org-string">'APOE'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/twas.org/apoe-fqtl-l.png" alt="apoe-fqtl-l.png">
</p>
</div>

<p>
Align the <code>fqtl</code> model and the GWAS summary statistics.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">B</span> = B.loc[twas_data[<span class="org-string">'3_y'</span>]]
<span class="org-variable-name">B</span> = B.loc[~B.index.duplicated(keep=<span class="org-string">'first'</span>)]
<span class="org-variable-name">snp_lodds</span> = snp_lodds.align(B, axis=0, join=<span class="org-string">'inner'</span>)[0]
<span class="org-variable-name">z</span> = twas_data[<span class="org-string">'5_x'</span>]
<span class="org-variable-name">z</span> = z.loc[~z.index.duplicated(keep=<span class="org-string">'first'</span>)]
<span class="org-variable-name">X</span> = x.loc[:,twas_data[<span class="org-string">'3_y'</span>]]
<span class="org-variable-name">X</span> = X.loc[:,~X.columns.duplicated(keep=<span class="org-string">'first'</span>)]
<span class="org-variable-name">u</span>, <span class="org-variable-name">d</span>, <span class="org-variable-name">vt</span> = np.linalg.svd(X, full_matrices=<span class="org-constant">False</span>)
<span class="org-variable-name">d</span>[d &lt; 1e-4] = 0
<span class="org-variable-name">R</span> = pd.DataFrame(vt.T.dot(np.diag(np.square(d))).dot(vt) / X.shape[0], index=X.columns, columns=X.columns)
</pre>
</div>

<p>
Compute the PIP for the tissue-specific effects.
</p>

<p>
\[ \Pr(b_{jt} \neq 0 \mid \cdot) = \sum_k \Pr(l_{jk} \neq 0 \mid \cdot) \Pr(f_{tk} \neq 0 \mid \cdot) \]
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">pip</span> = sp.expit(snp_lodds.loc[:,keep]) @ sp.expit(tis_lodds.loc[:,keep]).T
</pre>
</div>

<p>
Plot the permutation null distribution of the TWAS statistic.
</p>

\begin{align*}
  \vy &\sim \N(\veta a, \sigma^2 \mi), \qquad \veta = \mx\vb\\
  \hat{a} &= (\veta' \veta)^{-1} \veta' \vy\\
  &= (\vb' \mx' \mx \vb)^{-1} \vb' \mx' \vy\\
  &= \frac{\vb' \vz}{\vb' \mr \vb}, \qquad \vz = \mx' \vy / n, \mr = \mx' \mx / n
\end{align*}

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
<span class="org-variable-name">n_perm</span> = <span class="org-builtin">int</span>(1e6)
<span class="org-variable-name">pip_thresh</span> = 0.95
<span class="org-comment-delimiter"># </span><span class="org-comment">Hard threshold posterior mean B</span>
<span class="org-variable-name">B_thresh</span> = np.where(pip &gt; pip_thresh, B, 0)
<span class="org-variable-name">stat</span> = pd.Series(z @ B_thresh / (np.diag(B_thresh.T @ R @ B_thresh) + 1e-8), index=pip.columns)
<span class="org-variable-name">perm_stat</span> = []
<span class="org-comment-delimiter"># </span><span class="org-comment">Permute b_t by selecting new indices</span>
<span class="org-variable-name">n_qtls</span> = (pip &gt; pip_thresh).<span class="org-builtin">sum</span>(axis=0)
<span class="org-keyword">for</span> t <span class="org-keyword">in</span> B:
  <span class="org-keyword">if</span> n_qtls[t] == 0:
    perm_stat.append(np.repeat(0, n_perm))
    <span class="org-keyword">continue</span>
  <span class="org-variable-name">data</span> = np.repeat(B[t][pip[t] &gt; pip_thresh], n_perm)
  <span class="org-variable-name">row</span> = np.random.choice(B.shape[0], replace=<span class="org-constant">True</span>, size=(n_qtls[t], n_perm)).ravel()
  <span class="org-variable-name">col</span> = np.repeat(np.arange(n_perm), n_qtls[t])
  <span class="org-variable-name">bt</span> = ss.coo_matrix((data, (row, col)), shape=(B.shape[0], n_perm)).tocsr()
  perm_stat.append(z.values.reshape(1, -1) @ bt)
<span class="org-variable-name">perm_stat</span> = np.vstack(perm_stat)
<span class="org-variable-name">perm_logp</span> = -np.log10((np.square(stat.values).reshape(-1, 1) &lt;= np.square(perm_stat)).mean(axis=1) + 1e-8)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">tis</span> = [<span class="org-string">'Whole_Blood'</span>, <span class="org-string">'Lung'</span>, <span class="org-string">'Skin_Sun_Exposed_Lower_leg'</span>, <span class="org-string">'Brain_Caudate_basal_ganglia'</span>]
<span class="org-variable-name">idx</span> = stat.index.isin(tis)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 4)
fig.set_size_inches(8, 2.25)

<span class="org-keyword">for</span> a, s, p, t <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax, stat.loc[idx], perm_stat[idx].T, stat.index[idx]):
  a.hist(p, density=<span class="org-constant">True</span>, color=<span class="org-string">'0.7'</span>)
  a.axvline(s, lw=1, c=<span class="org-string">'r'</span>)
  a.set_title(<span class="org-string">' '</span>.join(t.split(<span class="org-string">'_'</span>)[:2]))
  a.set_xlabel(<span class="org-string">'TWAS statistic'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/twas.org/apoe-twas-null.png" alt="apoe-twas-null.png">
</p>
</div>

<p>
Implement adaptive permutation (<a href="http://www.biodatamining.org/content/7/1/9">Che et al. 2014</a>).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">max_success</span> = 10
<span class="org-variable-name">block_size</span> = <span class="org-builtin">int</span>(1e3)
<span class="org-variable-name">max_perm</span> = <span class="org-builtin">int</span>(1e7)
<span class="org-variable-name">n</span> = X.shape[0]
<span class="org-variable-name">pip_thresh</span> = 0.95
<span class="org-variable-name">B_thresh</span> = pd.DataFrame(np.where(pip &gt; pip_thresh, B, 0), index=B.index, columns=B.columns)
<span class="org-variable-name">n_qtls</span> = (pip &gt; pip_thresh).<span class="org-builtin">sum</span>(axis=0)
<span class="org-variable-name">num_success</span> = 0
<span class="org-variable-name">num_perm</span> = 0
<span class="org-variable-name">pvals</span> = []
<span class="org-keyword">for</span> t <span class="org-keyword">in</span> B:
  <span class="org-keyword">if</span> n_qtls[t] == 0:
    pvals.append(1)
    <span class="org-keyword">continue</span>
  <span class="org-variable-name">stat</span> = z @ B_thresh[t] / B_thresh[t].T @ R @ B_thresh[t]
  <span class="org-keyword">for</span> _ <span class="org-keyword">in</span> <span class="org-builtin">range</span>(max_perm // block_size):
    <span class="org-variable-name">perm_stat</span> = []
    <span class="org-variable-name">data</span> = np.repeat(B_thresh[t][pip[t] &gt; pip_thresh], block_size)
    <span class="org-variable-name">row</span> = np.random.choice(B_thresh.shape[0], replace=<span class="org-constant">True</span>, size=(n_qtls[t], block_size)).ravel()
    <span class="org-variable-name">col</span> = np.repeat(np.arange(block_size), n_qtls[t])
    <span class="org-variable-name">bt</span> = ss.coo_matrix((data, (row, col)), shape=(B_thresh.shape[0], block_size)).tocsr()
    <span class="org-variable-name">num</span> = (z.values.reshape(1, -1) @ bt)
    <span class="org-variable-name">den</span> = (R.values @ bt.power(2)).<span class="org-builtin">sum</span>(axis=0, keepdims=<span class="org-constant">True</span>)
    perm_stat.append(num / den)
    <span class="org-variable-name">perm_stat</span> = np.vstack(perm_stat)

    <span class="org-variable-name">num_success</span> += (np.square(perm_stat) &gt;= np.square(stat)).<span class="org-builtin">sum</span>()
    <span class="org-variable-name">num_perm</span> += block_size
    <span class="org-keyword">if</span> num_success &gt;= max_success:
      pvals.append(num_success / num_perm)
      <span class="org-keyword">break</span>
  <span class="org-keyword">else</span>:
    pvals.append((num_success + 1) / (max_perm + 1))
<span class="org-variable-name">pvals</span> = np.array(pvals)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(8, 6)
<span class="org-variable-name">grid</span> = np.arange(B.shape[1])
ax[0].bar(grid, z @ B_thresh / (np.diag(B_thresh.T @ R @ B_thresh) + 1e-8), width=.9, color=<span class="org-string">'k'</span>, zorder=3)
ax[0].set_ylabel(<span class="org-string">'TWAS statistic'</span>)
ax[1].bar(grid, -np.log(pvals), width=.9, color=<span class="org-string">'k'</span>, zorder=3)
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.set_xticks(grid)
  a.axhline(y=0, lw=1, color=<span class="org-string">'0.7'</span>)
  a.grid(axis=<span class="org-string">'x'</span>, color=<span class="org-string">'0.7'</span>)
ax[0].set_title(<span class="org-string">'APOE'</span>)
ax[1].set_xticklabels(B.columns, rotation=90)
ax[1].set_xlabel(<span class="org-string">'Tissue'</span>)
ax[1].set_ylabel(r<span class="org-string">'$-\log_{10}(p)$'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/twas.org/apoe.png" alt="apoe.png">
</p>
</div>

<p>
We expect an association between <i>APOE</i> gene expression in brain and AD;
however, we also find significant associations in several other
tissues. These associations appear to be driven by coding variants in <i>APOE</i>
which are predicted to have an effect in tissues other than brain. This
result could be expected, following the argument that strong effects on gene
expression of disease relevant tissues should not be tolerated by natural
selection, while strong effects in irrelevant tissues could be tolerated
(<a href="https://www.biorxiv.org/content/10.1101/459123v1">Wang et al. 2018</a>).
</p>
</div>
</div>

<div id="outline-container-org7afd361" class="outline-3">
<h3 id="org7afd361">Run TWAS</h3>
<div class="outline-text-3" id="text-org7afd361">
<p>
Run one chunk of the pipeline.
</p>

<div class="org-src-container">
<pre class="src src-shell">mkdir -p tasks
awk <span class="org-string">'{print &gt;"/broad/hptmp/aksarkar/tasks/chunk-" int((NR - 1) / 1000) ".txt"}'</span> geno/manifest
</pre>
</div>

<div class="org-src-container">
<pre class="src src-shell">qsub -cwd -V -terse -j y -sync n -N test -l <span class="org-variable-name">h_vmem</span>=1G,<span class="org-variable-name">h_rt</span>=0:30:00
<span class="org-builtin">source</span> activate /broad/compbio/aksarkar/.conda/envs/fqtl 
python /broad/compbio/aksarkar/projects/gtex-fqtl/twas.py /broad/hptmp/aksarkar/gtex-gwas-hg19/imputed_IGAP_Alzheimer.bed.gz /broad/hptmp/aksarkar/tasks/test.txt /broad/hptmp/aksarkar/twas/test
</pre>
</div>

<p>
Run the entire pipeline.
</p>

<div class="org-src-container">
<pre class="src src-shell">find /broad/hptmp/aksarkar/gtex-gwas-hg19 -name <span class="org-string">"*.bed.gz"</span> &gt;/broad/hptmp/aksarkar/gtex-gwas-hg19/manifest
parallel --dry-run <span class="org-string">'python /broad/compbio/aksarkar/projects/gtex-fqtl/twas.py {1} /broad/hptmp/aksarkar/tasks/chunk-{2}.txt /broad/hptmp/aksarkar/twas/{=1 use File::Basename; $_ = basename($_); s/.bed.gz//=}-{2}'</span> :::: /broad/hptmp/aksarkar/gtex-gwas-hg19/manifest ::: $(<span class="org-sh-quoted-exec">seq</span> 0 28) &gt;/broad/hptmp/aksarkar/twas/joblist
</pre>
</div>

<div class="org-src-container">
<pre class="src src-shell">qsub -cwd -V -terse -j y -sync n -N twas -S /bin/bash -l <span class="org-variable-name">h_vmem</span>=4G,<span class="org-variable-name">h_rt</span>=4:00:00 -t 1
<span class="org-builtin">readarray</span> -O1 tasks &lt; /broad/hptmp/aksarkar/twas/redo
<span class="org-builtin">source</span> activate /broad/compbio/aksarkar/.conda/envs/fqtl 
<span class="org-keyword">exec</span> ${<span class="org-variable-name">tasks</span>[$<span class="org-variable-name">SGE_TASK_ID</span>]}
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-04-27 Mon 16:17</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
